### PR TITLE
Refactor helm template and operator logic

### PR DESCRIPTION
**Pull request by vankichi**
### Description:

This PR includes refactoring benchmark operator logic and adding some new benchmark specs.

**Change benchmark operator logic**
 - Enable delete job pods when updating ValdBenchmarkJobResource
   - Due to the operator creating job resources using `controller-runtime`, the job pod is set to `blockOwnerDeletion=true` in owner reference. To enable deleting job pods when job resources are deleted, I have added the `DeletePropagationBackground` to the deleting option.

**Adding setting parameters**
To be a more customizable continuous benchmark tool, I have added the new parameters as followings:
- `enable_linear_search`: user can select whether to use a linear search process to calculate search recall when executing the search operation. linear search operation may require much time for execution, so it should be set as `false` when a user wants only to track search operation latency. 
- `ttl_seconds_after_finished`: user can set `ttl_seconds_after_finished` for each benchmark job resource. Completed job pods will be deleted when this value is set. On the other hand, pods will remain when it is not set.
- `dataset.url`: When the dataset URL is set and `dataset.name` is `original`, user can execute benchmark jobs with a user-defined HDF5 dataset, which satisfies the defined dataset group.
- `concurrencyLimit`: It controls the goroutine limit for executing the benchmark job. When the number of requests is large, it takes much time and requires CPU/Memory resources to complete all goroutines. I added this setting to save OOMKill or some failure based on resource failure.
- `search_config.aggregation_algorithm`: `aggregation_algorithm` is the new feature enabled from v1.7.6.

**Enable user-defined benchmark job image for  each operator**
Allowing user-defined benchmark job image for each benchmark operator by setting `job_image` at the `vald-benchmark-operator/values.yaml` 

### Versions:

- Go Version: 1.20.3
- Docker Version: 20.10.8
- Kubernetes Version: 1.22.0
- NGT Version: 2.0.11

### Checklist:

- [x] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/main/CONTRIBUTING.md) document and completed [our CLA agreement](https://cla-assistant.io/vdaas/vald).
- [x] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?

### CODE REVIEW

File: pkg/tools/benchmark/job/service/insert.go [hunk 1]
```diff
@@ -19,58 +19,73 @@ package service
...
-		if err != nil {
-			select {
-			case <-ctx.Done():
+	eg, egctx := errgroup.New(ctx)
+	eg.Limitation(j.concurrencyLimit)
+	for i := j.dataset.Range.Start; i <= j.dataset.Range.End; i++ {
+		iter := i
+		eg.Go(func() error {
+			log.Debugf("[benchmark job] Start insert: iter = %d", iter)
+			err := j.limiter.Wait(egctx)
```
Comment (hlts2) [inline pkg/tools/benchmark/job/service/insert.go:47]: I am not sure, but is it better the `j.limiter.Wait` function is called before the goroutine is generated? I think most of the goroutines will be in a waiting state.
```suggestion
	for i := j.dataset.Range.Start; i <= j.dataset.Range.End; i++ {
		iter := i
		err := j.limiter.Wait(egctx)
		if err != nil {
		
		}
		eg.Go(func() error {
			log.Debugf("[benchmark job] Start insert: iter = %d", iter)
```

File: internal/k8s/vald/benchmark/job/job_template_option.go [hunk 2]
```diff
@@ -17,21 +17,53 @@
...
 var defaultBenchmarkJobOpts = []BenchmarkJobOption{
 	WithSvcAccountName(SvcAccountName),
 	WithRestartPolicy(RestartPolicyNever),
 }
 
+// WithImage sets the docker image path for benchmark job.
+func WithImage(name string) BenchmarkJobOption {
+	return func(_ *jobs.Job) error {
+		if len(name) > 0 {
+			ContainerImage = name
```
[the remark this answers is not in the document]
Comment (vankichi) [reply internal/k8s/vald/benchmark/job/job_template_option.go]: @hlts2, thanks for reviewing!
I have fixed at https://github.com/vdaas/vald/pull/2043/commits/57ec0428556bb48dfa5b558d6d59e0a09f73d6db

File: pkg/tools/benchmark/job/service/insert.go [hunk 3]
```diff
@@ -19,58 +19,73 @@ package service
...
+				case ech <- err:
 				}
-			default:
-				st, _ := status.FromError(err)
-				log.Warnf("[benchmark job] insert error is detected: code = %d, msg = %s", st.Code(), err.Error())
 			}
-		}
-		// TODO: send metrics to the Prometeus
-		log.Infof("[benchmark job] Finish insert: iter= %d \n%v\n", i, res)
+			loop_cnt := math.Floor(float64(i-1) / float64(len(vecs)))
```
Comment (hlts2) [inline pkg/tools/benchmark/job/service/insert.go:60]:
```suggestion
			loopCnt := math.Floor(float64(i-1) / float64(len(vecs)))
```

File: pkg/tools/benchmark/job/service/job.go [hunk 4]
```diff
@@ -186,20 +186,65 @@ func New(opts ...Option) (Job, error) {
...
+		log.Infof("[benchmark job] start load dataset of %s", j.hdf5.GetName().String())
+		var key hdf5.Hdf5Key
+		switch j.dataset.Group {
+		case "train":
+			key = hdf5.Train
+		case "test":
+			key = hdf5.Test
+		case "neighbors":
+			key = hdf5.Neighors
+		default:
```
Comment (hlts2) [inline pkg/tools/benchmark/job/service/job.go:213]: Is it better to return an error when the dataset group is unknown?

File: pkg/tools/benchmark/job/service/search.go [hunk 5]
```diff
@@ -19,98 +19,150 @@ package service
...
-				st, _ := status.FromError(err)
-				if st.Code() != codes.NotFound {
-					log.Warnf("[benchmark job] search error is detected: code = %d, msg = %s", st.Code(), err.Error())
+				case <-egctx.Done():
+					return egctx.Err()
+				case ech <- err:
 				}
 			}
+			iter := i
+			loop_cnt := math.Floor(float64(i-1) / float64(len(vecs)))
```
Comment (hlts2) [inline pkg/tools/benchmark/job/service/search.go:121]:
```suggestion
			loopCnt := math.Floor(float64(i-1) / float64(len(vecs)))
```

File: pkg/tools/benchmark/job/service/update.go [hunk 6]
```diff
@@ -19,60 +19,76 @@ package service
...
+					// return errors.Join(err, context.Canceled)
+					return nil
+				}
+				select {
+				case <-egctx.Done():
+					return egctx.Err()
+				case ech <- err:
 				}
+			}
+			loop_cnt := math.Floor(float64(i-1) / float64(len(vecs)))
```
Comment (hlts2) [inline pkg/tools/benchmark/job/service/update.go:61]:
```suggestion
			loopCnt := math.Floor(float64(i-1) / float64(len(vecs)))
```

File: pkg/tools/benchmark/job/service/insert.go [hunk 7]
```diff
@@ -19,58 +19,73 @@ package service
...
+			idx := i - 1 - (len(vecs) * int(loop_cnt))
```
Comment (hlts2) [inline pkg/tools/benchmark/job/service/insert.go:61]:
```suggestion
			idx := iter - 1 - (len(vecs) * int(loop_cnt))
```

File: pkg/tools/benchmark/job/service/insert.go (hunk 3 above)
Comment (hlts2) [inline pkg/tools/benchmark/job/service/insert.go:60]:
```suggestion
			loop_cnt := math.Floor(float64(iter-1) / float64(len(vecs)))
```

**Review by hlts2 — APPROVED** [review]

**Pull request merged**